sql/src/main/java/io/crate/planner/consumer/ManyTableConsumer.java

```diff
@@ -89,99 +90,195 @@ public Plan consume(AnalyzedRelation relation, ConsumerContext context) {
-        outerloop:
-        for (List<QualifiedName> permutation : Collections2.permutations(relations)) {
-            if (!preSorted.equals(permutation.subList(0, preSorted.size()))) {
-                continue;
-            }
-            currentPermutationJoinPairs.clear();
-            int joinPushDowns = 0;
-            for (int i = 0; i < permutation.size() - 1; i++) {
-                QualifiedName a = permutation.get(i);
-                QualifiedName b = permutation.get(i + 1);
+        LinkedHashSet<QualifiedName> bestOrder = new LinkedHashSet<>();
+        List<Set<QualifiedName>> sets = new ArrayList<>(explicitJoinedRelations);
+        sets.addAll(implicitJoinedRelations);
+
+        // If no `ORDER BY` present we have no preSort to follow so we return the relations in ordering
+        // obtained by the join conditions (explicit and/or implicit) between them.
+        if (preSorted.isEmpty()) {
+            ObjectIntHashMap<QualifiedName> occurrences =
+                getOccurrencesInJoinConditions(relations, explicitJoinedRelations, implicitJoinedRelations);
+
+            Set<QualifiedName> firstJoinPair = getFirstJoinPair(occurrences, sets);
+            assert firstJoinPair != null : "firstJoinPair should not be null";
+            bestOrder.addAll(firstJoinPair);
+        } else {
+            bestOrder.addAll(preSorted);
+        }
 
-                JoinPair joinPair = JoinPairs.ofRelations(a, b, joinPairs, false);
-                if (joinPair == null) {
-                    // relations are not directly joined, lets check if they are part of an outer join
-                    if (outerJoinRelations.contains(a) || outerJoinRelations.contains(b)) {
-                        // part of an outer join, don't change pairs, permutation not possible
-                        continue outerloop;
-                    } else {
-                        pair.clear();
-                        pair.add(a);
-                        pair.add(b);
-                        joinPushDowns += implicitJoinedRelations.contains(pair) ? 1 : 0;
-                        currentPermutationJoinPairs.add(JoinPair.crossJoin(a, b));
-                    }
-                } else {
-                    if (JoinPairs.joinConditionIncludesRelations(currentPermutationJoinPairs, joinPair)) {
-                        joinPushDowns += 1;
-                    }
-                    currentPermutationJoinPairs.add(JoinPair.crossJoin(a, b));
+        buildBestOrderByJoinConditions(sets, bestOrder);
+        bestOrder.addAll(relations);
+        return bestOrder;
+    }
+
+    private static ObjectIntHashMap<QualifiedName> getOccurrencesInJoinConditions(
+        Collection<QualifiedName> relations,
+        Set<? extends Set<QualifiedName>> explicitJoinedRelations,
+        Set<? extends Set<QualifiedName>> implicitJoinedRelations) {
+
+        List<QualifiedName> qualifiedNames = new ArrayList<>(explicitJoinedRelations.size() + implicitJoinedRelations.size());
```

Comment:
Looks like it would be possible to build the occurrences map below without creating this qualifiedNames list by looping over the explicit/implicit relations directly?